File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/foundation/bridge/ComponentVTable.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * ComponentVTable.kt
 *
 * Per-modality vtable describing the 5 ops that vary across the
 * LLM / STT / TTS / VAD / VLM component actor scaffolds. The rest of
 * the actor (handle caching, error wrapping, lifecycle gates) is
 * generic and will live in a shared ComponentActor scaffold (Kotlin
 * counterpart of Swift's CppBridge.ComponentActor) introduced in
 * subsequent waves.
 *
 * VoiceAgent is intentionally NOT modeled here — its handle type wraps
 * a composite (STT + LLM + TTS + VAD) and create() is async. VoiceAgent
 * keeps its own bespoke scaffold (mirrors Swift's exception).
 *
 * Kotlin SDK mirror of Swift's
 * `Sources/RunAnywhere/Foundation/Bridge/ComponentVTable.swift`.
 * Design (Option A): expect/class with a companion object that
 * exposes the 5 modality instances (`llm`, `stt`, `tts`, `vad`, `vlm`).
 * Platform implementations wire to RunAnywhereBridge JNI primitives.
 */

package com.runanywhere.sdk.foundation.bridge

import ai.runanywhere.proto.v1.SDKComponent
import com.runanywhere.sdk.native.bridge.RunAnywhereBridge

/**
 * jvmAndroid actual. See the commonMain `expect` declaration for the
 * shared contract.
 */
public class ComponentVTable internal constructor(
    public val component: SDKComponent,
    private val createFn: () -> Long,
    private val isLoadedFn: (Long) -> Boolean,
    private val destroyFn: (Long) -> Unit,
    private val cleanupFn: (Long) -> Int,
    public val loadModel: ((handle: Long, path: String, id: String, name: String) -> Int)?,
) {
    public fun create(): Long = createFn()

    public fun destroy(handle: Long) = destroyFn(handle)

    public fun isLoaded(handle: Long): Boolean = isLoadedFn(handle)

    public fun cleanup(handle: Long) {
        cleanupFn(handle)
    }

    public companion object {
        public val llm: ComponentVTable =
            ComponentVTable(
                component = SDKComponent.SDK_COMPONENT_LLM,
                createFn = { RunAnywhereBridge.racLlmComponentCreate() },
                isLoadedFn = { handle -> RunAnywhereBridge.racLlmComponentIsLoaded(handle) },
                destroyFn = { handle -> RunAnywhereBridge.racLlmComponentDestroy(handle) },
                cleanupFn = { handle -> RunAnywhereBridge.racLlmComponentCleanup(handle) },
                loadModel = { handle, path, id, name -> RunAnywhereBridge.racLlmComponentLoadModel(handle, path, id, name) },
            )

        public val stt: ComponentVTable =
            ComponentVTable(
                component = SDKComponent.SDK_COMPONENT_STT,
                createFn = { RunAnywhereBridge.racSttComponentCreate() },
                isLoadedFn = { handle -> RunAnywhereBridge.racSttComponentIsLoaded(handle) },
                destroyFn = { handle -> RunAnywhereBridge.racSttComponentDestroy(handle) },
                cleanupFn = { handle -> RunAnywhereBridge.racSttComponentCleanup(handle) },
                loadModel = { handle, path, id, name -> RunAnywhereBridge.racSttComponentLoadModel(handle, path, id, name) },
            )

        public val tts: ComponentVTable =
            ComponentVTable(
                component = SDKComponent.SDK_COMPONENT_TTS,
                createFn = { RunAnywhereBridge.racTtsComponentCreate() },
                isLoadedFn = { handle -> RunAnywhereBridge.racTtsComponentIsLoaded(handle) },
                destroyFn = { handle -> RunAnywhereBridge.racTtsComponentDestroy(handle) },
                cleanupFn = { handle -> RunAnywhereBridge.racTtsComponentCleanup(handle) },
                loadModel = { handle, path, id, name -> RunAnywhereBridge.racTtsComponentLoadVoice(handle, path, id, name) },
            )

        public val vad: ComponentVTable =
            ComponentVTable(
                component = SDKComponent.SDK_COMPONENT_VAD,
                createFn = { RunAnywhereBridge.racVadComponentCreate() },
                isLoadedFn = { handle -> RunAnywhereBridge.racVadComponentIsLoaded(handle) },
                destroyFn = { handle -> RunAnywhereBridge.racVadComponentDestroy(handle) },
                cleanupFn = { handle -> RunAnywhereBridge.racVadComponentCleanup(handle) },
                loadModel = { handle, path, id, name -> RunAnywhereBridge.racVadComponentLoadModel(handle, path, id, name) },
            )
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Foundation/Bridge/ComponentVTable.swift
//
//  ComponentVTable.swift
//  RunAnywhere SDK
//
//  Per-modality vtable describing the 5 ops that vary across the
//  LLM / STT / TTS / VAD / VLM component actor scaffolds. The rest of
//  the actor (handle caching, error wrapping, lifecycle gates) is
//  generic and lives in `CppBridge.ComponentActor`.
//
//  VoiceAgent is intentionally NOT modeled here — its handle type
//  is `rac_voice_agent_handle_t` (a struct pointer), and `create` is
//  async + composite (it must gather LLM/STT/TTS/VAD handles first).
//  VoiceAgent keeps its own bespoke actor scaffold.
//
//  This collapses ~885 LOC of duplicated boilerplate.
//

import CRACommons

extension CppBridge {

    /// Function-pointer table parameterizing one modality's component actor.
    ///
    /// Five ops vary per modality; the actor scaffold (lazy create,
    /// `isLoaded`, `loadModel`, `unload`, `destroy`) is shared.
    public struct ComponentVTable: Sendable {

        // MARK: - Identity

        /// The proto-canonical component identity. Used for log/error labels
        /// and for keying lifecycle state across components.
        public let component: RASDKComponent

        // MARK: - Lifecycle ops

        /// Create the C++ component, writing the new opaque handle into `out`.
        public let create: @Sendable (_ out: UnsafeMutablePointer<rac_handle_t?>) -> rac_result_t

        /// Query whether the component has a model/voice currently loaded.
        public let isLoaded: @Sendable (_ handle: rac_handle_t) -> rac_bool_t

        /// Cleanup (unload) the loaded asset, leaving the component reusable.
        public let cleanup: @Sendable (_ handle: rac_handle_t) -> Void

        /// Destroy the underlying C++ component and release its resources.
        public let destroy: @Sendable (_ handle: rac_handle_t) -> Void

        /// Load a model/voice given (path, id, name). `nil` means the modality
        /// has no path-based load (e.g. richer custom signatures live on the
        /// per-modality extension). Stays optional so VLM (extra projector
        /// path) can opt out without adding a third arg shape here.
        public let loadModel: (@Sendable (
            _ handle: rac_handle_t,
            _ path: UnsafePointer<CChar>?,
            _ id: UnsafePointer<CChar>?,
            _ name: UnsafePointer<CChar>?
        ) -> rac_result_t)?
    }
}

// MARK: - Static vtable instances

extension CppBridge.ComponentVTable {

    /// LLM component vtable — `rac_llm_component_*` family.
    public static let llm = CppBridge.ComponentVTable(
        component: .llm,
        create: { rac_llm_component_create($0) },
        isLoaded: { rac_llm_component_is_loaded($0) },
        cleanup: { _ = rac_llm_component_cleanup($0) },
        destroy: { rac_llm_component_destroy($0) },
        loadModel: { handle, path, id, name in
            rac_llm_component_load_model(handle, path, id, name)
        }
    )

    /// STT component vtable — `rac_stt_component_*` family.
    public static let stt = CppBridge.ComponentVTable(
        component: .stt,
        create: { rac_stt_component_create($0) },
        isLoaded: { rac_stt_component_is_loaded($0) },
        cleanup: { _ = rac_stt_component_cleanup($0) },
        destroy: { rac_stt_component_destroy($0) },
        loadModel: { handle, path, id, name in
            rac_stt_component_load_model(handle, path, id, name)
        }
    )

    /// TTS component vtable — `rac_tts_component_*` family.
    /// The "model" generic name aliases TTS's "voice" terminology at the C ABI.
    public static let tts = CppBridge.ComponentVTable(
        component: .tts,
        create: { rac_tts_component_create($0) },
        isLoaded: { rac_tts_component_is_loaded($0) },
        cleanup: { _ = rac_tts_component_cleanup($0) },
        destroy: { rac_tts_component_destroy($0) },
        loadModel: { handle, path, id, name in
            rac_tts_component_load_voice(handle, path, id, name)
        }
    )

    /// VAD component vtable — `rac_vad_component_*` family.
    public static let vad = CppBridge.ComponentVTable(
        component: .vad,
        create: { rac_vad_component_create($0) },
        isLoaded: { rac_vad_component_is_loaded($0) },
        cleanup: { _ = rac_vad_component_cleanup($0) },
        destroy: { rac_vad_component_destroy($0) },
        loadModel: { handle, path, id, name in
            rac_vad_component_load_model(handle, path, id, name)
        }
    )

    /// VLM component vtable — `rac_vlm_component_*` family.
    ///
    /// SDK-side `loadModel(from:)` adapters were removed from the
    /// VLM actor; the level-3 handle is never loaded with a model and the
    /// `loadModel` slot is dead in practice. The slot is kept here only so
    /// the `ComponentVTable` shape stays uniform with the sibling modalities
    /// (LLM, STT, TTS, VAD). Inference and cancel now route through the
    /// lifecycle service via the proto ABI.
    public static let vlm = CppBridge.ComponentVTable(
        component: .vlm,
        create: { rac_vlm_component_create($0) },
        isLoaded: { rac_vlm_component_is_loaded($0) },
        cleanup: { _ = rac_vlm_component_cleanup($0) },
        destroy: { rac_vlm_component_destroy($0) },
        loadModel: { handle, path, id, name in
            // Passing nil for vision_projector_path covers single-artifact
            // VLMs; the multi-artifact VLM path stays on the modality
            // extension where it can pass the projector explicitly.
            rac_vlm_component_load_model(handle, path, nil, id, name)
        }
    )
}
